src/mesh/region_set.cc

```diff
@@ -0,0 +1,385 @@
+			if (it_map != el_to_reg_map_.end()) {
+				xprintf(Warn, "Region assigned to element with id %u will be rewritten.\n", (*it_element));
+			}
+			el_to_reg_map_.insert( std::make_pair((*it_element), region_id) );
+
+		}
+	}
+}
+
+
+
+const IT::Record & RegionSetFromElements::get_region_input_type()
+{
+    return IT::Record("From_Elements", "Region declared by name, ID and enum of elements.\n"
+    								   "Allows to get existing region or create new and assign\n"
+    								   "elements to its. Elements are specified by ids.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"Label (name) of the region. Has to be unique in one mesh.\n")
+		.declare_key("id", IT::Integer(0), IT::Default::optional(),
+				"The ID of the region to which you assign label.\n"
+				"If new region is created and ID is not set, unique ID will be generated automatically.")
+		.declare_key("element_list", IT::Array( IT::Integer(0) ), IT::Default::optional(),
+				"Specification of the region by the list of elements. This is not recommended")
+		.close();
+}
+
+
+
+const int RegionSetFromElements::registrar =
+		Input::register_class< RegionSetFromElements, const Input::Record &, Mesh * >("From_Elements") +
+		RegionSetFromElements::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetBoundary
+ * Need new implementation, will be solved later.
+ */
+
+// RegionSetBoundary::RegionSetBoundary(const Input::Record &rec, Mesh *mesh) : RegionSetBase(mesh) {}
+
+// const IT::Record & RegionSetBoundary::get_region_input_type() {}
+
+//const int RegionSetBoundary::registrar =
+//		Input::register_class< RegionSetBoundary, const Input::Record &, Mesh * >("Region_Boundary") +
+//		RegionSetBoundary::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetUnion
+ */
+
+RegionSetUnion::RegionSetUnion(const Input::Record &rec, Mesh *mesh)
+: RegionSetBase(mesh)
+{
+	string name_of_set = rec.val<string>("name");
+	Input::Iterator<Input::Array> region_ids = rec.find<Input::Array>("region_ids");
+	Input::Iterator<Input::Array> regions = rec.find<Input::Array>("regions");
+
+	std::set<Region, bool (*)(const Region&, const Region&)> set(Region::comp);
+	if (region_ids) {
+		for (Input::Iterator<unsigned int> it_ids = region_ids->begin<unsigned int>();
+				it_ids != region_ids->end();
+		        ++it_ids) {
+			try {
+				Region reg = region_db_.find_id(*it_ids);
+				if (reg.is_valid()) {
+					set.insert(reg); // add region if doesn't exist in set
+				} else {
+					xprintf(Warn, "Region with id %d doesn't exist. Skipping\n", (*it_ids));
+				}
+			} catch(RegionDB::ExcUniqueRegionId &e) {
+				e << region_ids->ei_address();
+				throw;
+			}
+		}
+	}
+
+	if (regions) {
```

Comment:
In order to reuse this code in update_history:
- make a union method in RegionDB getting vector of region names returning a region set.
- just convert IDs to labels for region_ids, merge that list with "regions" whole list pass to the new RegionDB::union method.